python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py

```diff
@@ -0,0 +1,211 @@
+from pprint import pformat
+from typing import Any, Dict, List, Optional, cast
+
+import boto3
+import botocore
+import dagster._check as check
+from dagster import PipesClient
+from dagster._annotations import experimental
+from dagster._core.definitions.resource_annotation import TreatAsResourceParam
+from dagster._core.errors import DagsterExecutionInterruptedError
+from dagster._core.execution.context.compute import OpExecutionContext
+from dagster._core.pipes.client import (
+    PipesClientCompletedInvocation,
+    PipesContextInjector,
+    PipesMessageReader,
+)
+from dagster._core.pipes.utils import PipesEnvContextInjector, open_pipes_session
+from mypy_boto3_ecs.client import ECSClient
+from mypy_boto3_ecs.type_defs import RunTaskRequestRequestTypeDef
+from typing_extensions import Unpack
+
+from dagster_aws.pipes.message_readers import PipesCloudWatchMessageReader
+
+
+@experimental
+class PipesECSClient(PipesClient, TreatAsResourceParam):
+    """A pipes client for running AWS ECS tasks.
+
+    Args:
+        client (Optional[boto3.client]): The boto ECS client used to launch the ECS task
+        context_injector (Optional[PipesContextInjector]): A context injector to use to inject
+            context into the ECS task. Defaults to :py:class:`PipesEnvContextInjector`.
+        message_reader (Optional[PipesMessageReader]): A message reader to use to read messages
+            from the ECS task. Defaults to :py:class:`PipesCloudWatchMessageReader`.
+        forward_termination (bool): Whether to cancel the ECS task when the Dagster process receives a termination signal.
+    """
+
+    def __init__(
+        self,
+        client: Optional[ECSClient] = None,
+        context_injector: Optional[PipesContextInjector] = None,
+        message_reader: Optional[PipesMessageReader] = None,
+        forward_termination: bool = True,
+    ):
+        self._client = client or boto3.client("ecs")
+        self._context_injector = context_injector or PipesEnvContextInjector()
+        self._message_reader = message_reader or PipesCloudWatchMessageReader()
+        self.forward_termination = check.bool_param(forward_termination, "forward_termination")
+
+    @classmethod
+    def _is_dagster_maintained(cls) -> bool:
+        return True
+
+    def run(
+        self,
+        *,
+        context: OpExecutionContext,
+        task_definition: str,
+        extras: Optional[Dict[str, Any]] = None,
+        **params: Unpack[RunTaskRequestRequestTypeDef],
```

Comment:
I'm not convinced that this should be kwarged.

I think this should be `ecs_task_params` and be a standalone argument.

Reply:
Do you have any rules to decide when to use kwargs vs a dict argument? I can't think about any major differences except that it would be a few extra characters to type with the dict. Also, kwargs feel more natural to me. But definitely dict works, we can do it. I'll change it to dict if you don't leave any comments in this thread.

Reply:
You have to be very confident that there will never be name collisions between `RunTaskRequestRequestTypeDef` the named arguments.  If you add an argument to run in the future that collides with a key `RunTaskRequestRequestTypeDef`, it will break. Similarly if ecs adds a new parameter that conflicts with something in run, it will all break.

Reply:
That's a good point. Let's do it this way.